#!/usr/bin/env python3

import grpc
import json
import time
import proto.chassis_controller_pb2 as chassis_pb2
import proto.chassis_controller_pb2_grpc as chassis_grpc
import proto.upper_controller_pb2 as upper_pb2
import proto.upper_controller_pb2_grpc as upper_grpc

def test_chassis_service():
    """æµ‹è¯•åº•ç›˜æœåŠ¡"""
    print("ğŸ”§ æµ‹è¯•åº•ç›˜gRPCæœåŠ¡...")
    try:
        channel = grpc.insecure_channel('localhost:50051')
        stub = chassis_grpc.ChassisControlerStub(channel)
        
        cmd = chassis_pb2.Command(
            linear=chassis_pb2.Descartes(x=1.0, y=0.0, z=0.0),
            angular=chassis_pb2.Descartes(x=0.0, y=0.0, z=0.0),
            tap=0,
            zOff=0.0
        )
        
        resp = stub.sendCommand(cmd)
        print(f"âœ… åº•ç›˜æœåŠ¡æµ‹è¯•æˆåŠŸ: {resp.msg}")
        return True
    except Exception as e:
        print(f"âŒ åº•ç›˜æœåŠ¡æµ‹è¯•å¤±è´¥: {e}")
        return False

def test_upper_service():
    """æµ‹è¯•ä¸Šè‚¢æœåŠ¡"""
    print("ğŸ”§ æµ‹è¯•ä¸Šè‚¢gRPCæœåŠ¡...")
    try:
        channel = grpc.insecure_channel('localhost:50052')
        stub = upper_grpc.UpperControllerStub(channel)
        
        payload = upper_pb2.ArmPayload(
            arm=upper_pb2.ArmPosition(left=[1.0, 2.0, 3.0], right=[4.0, 5.0, 6.0]),
            effector=upper_pb2.EffectorPosition(left=[0.1, 0.2], right=[0.3, 0.4])
        )
        
        resp = stub.sendArmAction(payload)
        print(f"âœ… ä¸Šè‚¢æœåŠ¡æµ‹è¯•æˆåŠŸ: {resp.msg}")
        return True
    except Exception as e:
        print(f"âŒ ä¸Šè‚¢æœåŠ¡æµ‹è¯•å¤±è´¥: {e}")
        return False

def test_workflow_messages():
    """æµ‹è¯•å·¥ä½œæµæ¶ˆæ¯æ ¼å¼"""
    print("ğŸ”§ æµ‹è¯•å·¥ä½œæµæ¶ˆæ¯æ ¼å¼...")
    
    # æµ‹è¯•åº•ç›˜å‘½ä»¤
    chassis_cmd = {
        "action": "MOVE",
        "target": {"x": 1.0, "y": 0.0, "z": 0.0, "wz": 0.0},
        "tap": 0,
        "zOff": 0.0
    }
    print(f"âœ… åº•ç›˜å‘½ä»¤æ ¼å¼: {json.dumps(chassis_cmd, indent=2)}")
    
    # æµ‹è¯•æœºæ¢°è‡‚å‘½ä»¤
    arm_cmd = {
        "action": "GRAB",
        "target": {"left": [1.0, 2.0, 3.0], "right": [4.0, 5.0, 6.0]},
        "effector": {"left": [0.1, 0.2], "right": [0.3, 0.4]}
    }
    print(f"âœ… æœºæ¢°è‡‚å‘½ä»¤æ ¼å¼: {json.dumps(arm_cmd, indent=2)}")
    
    # æµ‹è¯•çŠ¶æ€æ¶ˆæ¯
    status_msg = {
        "action": "MOVE_COMPLETE",
        "message": "åº•ç›˜ç§»åŠ¨å®Œæˆ"
    }
    print(f"âœ… çŠ¶æ€æ¶ˆæ¯æ ¼å¼: {json.dumps(status_msg, indent=2)}")
    
    return True

def main():
    """ä¸»æµ‹è¯•å‡½æ•°"""
    print("ğŸš€ å¼€å§‹æµ‹è¯•Doraæœºå™¨äººå·¥ä½œæµç»„ä»¶...")
    print("=" * 50)
    
    # æµ‹è¯•gRPCæœåŠ¡
    chassis_ok = test_chassis_service()
    upper_ok = test_upper_service()
    
    # æµ‹è¯•æ¶ˆæ¯æ ¼å¼
    msg_ok = test_workflow_messages()
    
    print("=" * 50)
    print("ğŸ“Š æµ‹è¯•ç»“æœæ±‡æ€»:")
    print(f"  åº•ç›˜æœåŠ¡: {'âœ… é€šè¿‡' if chassis_ok else 'âŒ å¤±è´¥'}")
    print(f"  ä¸Šè‚¢æœåŠ¡: {'âœ… é€šè¿‡' if upper_ok else 'âŒ å¤±è´¥'}")
    print(f"  æ¶ˆæ¯æ ¼å¼: {'âœ… é€šè¿‡' if msg_ok else 'âŒ å¤±è´¥'}")
    
    if chassis_ok and upper_ok and msg_ok:
        print("ğŸ‰ æ‰€æœ‰æµ‹è¯•é€šè¿‡ï¼å¯ä»¥å¯åŠ¨Doraå·¥ä½œæµäº†ã€‚")
        print("\nğŸ“‹ ä¸‹ä¸€æ­¥:")
        print("1. ç¡®ä¿gRPCæœåŠ¡æ­£åœ¨è¿è¡Œ")
        print("2. è¿è¡Œ: dora run dataflow.yml")
    else:
        print("âŒ éƒ¨åˆ†æµ‹è¯•å¤±è´¥ï¼Œè¯·æ£€æŸ¥æœåŠ¡çŠ¶æ€ã€‚")

if __name__ == "__main__":
    main() 